### PR TITLE
Fix uncaught exception reporting for Scheduler

### LINKED ISSUE #799: API for overriding Scheduler.reportException?
On JS, I wanted to override `reportException` in a Scheduler and ended up in a bad situation:

* `AsyncScheduler` and `TrampolineScheduler` are final, so I cannot inherit from any of them.
* There are no utilities like `withExecutionModel` for creating one.
* There are no factory methods on Scheduler that take an `UncaughtExceptionReporter` for JS (there are for JVM).

I think we can port a number of JVM-only constructors to JS, such as this one:
```scala
def apply(reporter: UncaughtExceptionReporter, executionModel: ExecModel): Scheduler
```
and those taking ExecutionContext (since we have ECs in SJS, and schedulers are ECs themselves)

I'd also appreciate a method like `withExceptionReporter` too.

### PR DESCRIPTION
**Pull request by oleg-py**
Fixes #799, #829.

### CODE REVIEW

File: monix-execution/shared/src/test/scala/monix/execution/schedulers/UncaughtExceptionReporterSuite.scala [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import monix.execution.{ExecutionModel, FutureUtils, Scheduler, UncaughtExceptionReporter}
+
+
+class UncaughtExceptionReporterBaseSuite extends TestSuite[Promise[Throwable]] {
+  object Dummy extends Throwable
+  private[this] val throwRunnable: Runnable = new Runnable {
+    def run(): Unit = throw Dummy
+  }
+
+  private[this] val immediateEC = new ExecutionContext {
```
Comment (Avasil) [inline monix-execution/shared/src/test/scala/monix/execution/schedulers/UncaughtExceptionReporterSuite.scala]: I think we could use `monix.execution.schedulers.TrampolineExecutionContext.immediate`

File: monix-execution/js/src/main/scala/monix/execution/schedulers/SchedulerCompanionImpl.scala [hunk 2]
```diff
@@ -32,9 +32,19 @@ private[execution] class SchedulerCompanionImpl extends SchedulerCompanion {
...
   def apply(
     context: ExecutionContext = StandardContext,
-    executionModel: ExecModel = ExecModel.Default): Scheduler =
+    executionModel: ExecModel = ExecModel.Default,
+  ): Scheduler =
     AsyncScheduler(context, executionModel)
 
+  def apply(
+    ec: ExecutionContext,
+    reporter: UncaughtExceptionReporter): Scheduler =
```
[the remark this answers is not in the document]
Comment (oleg-py) [reply monix-execution/js/src/main/scala/monix/execution/schedulers/SchedulerCompanionImpl.scala]: Thanks for nits, actually, it's very helpful 👍

**Comment by alexandru** [comment]
Hey guys, this is looking good, so what's left to do on it?

**Comment by Avasil** [comment]
@alexandru Oh, I meant to assign for review, to double check but if it's fine then we can merge it :)

**Pull request merged**